website/blog/2024-04-26-deep-learning-part-2/index.md

```diff
@@ -0,0 +1,287 @@
+trainer = Trainer(
+   callbacks=[
+       EarlyStopping(
+           monitor="Val_F1_Score",
+           min_delta=train_config.min_delta,
+           patience=train_config.patience,
+           verbose=True,
+           mode="max",
+       )
+   ],
+   default_root_dir=train_config.model_checkpoint_dir,
+   fast_dev_run=bool(train_config.debug_mode_sample),
+   max_epochs=train_config.max_epochs,
+   max_time=train_config.max_time,
+   precision="32-true"
+)
+
+trainer.fit(model=nlp_model, datamodule=datamodule)
+```
+
+## Visualization and Sharing Capabilities within MLflow
+
+The newly introduced DL-specific visualization capabilities introduced in MLflow 2.12 enable you to make comparisons between different runs and artifacts over epochs.
+When comparing training runs, MLflow is capable of generating useful visualization that can be integrated into dashboards, facilitating
+easy sharing. Additionally, the centralized storage of metrics, in conjunction with parameters, allows for effective analysis of the training
+efficacy, as shown in the image below.
+
+![Epoch Run Compare](compare.png)
+
+## When to stop training?
+
+When training DL models, it is important to understand when to stop. Efficient training (for minimizing the overall cost incurred for
+conducting training) and optimal model performance rely heavily on preventing a model from overfitting on the training data. A model
+that trains for too long will invariably become quite good at effectively ‘memorizing’ the training data, resulting in a reduction in
+the performance of the model when presented with novel data. A straightforward way to evaluate this behavior is to ensure that
+validation data set metrics (scoring loss metrics on data that is not in the training data set) are captured during the training
+loop. Integrating the MLflow callback into the PyTorch Lightning Trainer allows for iterative logging of loss metrics at
+configurable iterations, enabling an easily debuggable evaluation of the training performance, ensuring that stopping criteria
+can be enforced at the appropriate time to prevent overfitting.
+
+### Evaluating epoch checkpoints of Fine Tuned Models with MLflow
+
+With your training process meticulously tracked and logged by MLflow, you have the flexibility to retrieve and test your model at
+any arbitrary checkpoint. To do this, you can use the mlflow.pytorch.load_model() API to load the model from a specific run
+and use the `predict()` method for evaluation.
+
+In the example below, we will load the model checkpoint from the 3rd epoch and use the `Lightning` train module to generate predictions based on the
+checkpoint state of the saved training epoch.
+
+```python
+import mlflow
+
+
+mlflow.pytorch.autolog(disable = True)
+
+run_id = '<Add the run ID>'
+model = mlflow.pytorch.load_model(f"runs:/{run_id}/model")
+
+model = mlflow.pytorch.load_checkpoint(TransformerModule, run_id, 3)
+
+train_module.predict(model, dataloaders = DataLoader(ds))
```

Comment:
Do we need to do this instead of `model.predict()?`

Reply:
Yes. because the loaded object is a Pytorch object (not a pyfunc object) which does not have a predict method. 

`type(model)`
> custom_module.fine_tune_clsify_head.TransformerModule